### PR TITLE
YD-400 Phone number whitelist

### PR DESCRIPTION
**Pull request by aristotelos**
Tested manually with the property enabled

### CODE REVIEW

File: adminservice/src/main/resources/templates/whitelisted-numbers.vm [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+<html>
+<head><title>Whitelisted numbers</title></head>
+<body>
+	<h2>Add whitelisted number</h2>
+	<form action="" method="post">
+        <p><label>Number: <input type="tel" name="mobileNumber" title="e.g. +31612345678" placeholder="+31612345678" pattern="\+[1-9][0-9]+" /></label></p>
+        <p><input type="submit" value="Submit" /></p>
+    </form>
+    
+    <h2>Current whitelisted numbers</h2>
```
Comment (Bert-R) [inline adminservice/src/main/resources/templates/whitelisted-numbers.vm]: Change in one of:
* Currently white-listed numbers
* Current set of white-listed numbers

File: adminservice/src/main/java/nu/yona/server/subscriptions/rest/WhitelistedNumberController.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+
+import nu.yona.server.subscriptions.service.WhitelistedNumberService;
+
+@Controller
+@RequestMapping(value = "/whitelistedNumbers")
+public class WhitelistedNumberController
```
Comment (Bert-R) [inline adminservice/src/main/java/nu/yona/server/subscriptions/rest/WhitelistedNumberController.java]: White list and white-listed are spelled as two words (see [Merriam Webster](https://www.merriam-webster.com/dictionary/whitelist)). Given that this is temporary code, I'm OK leaving it in, except for user-facing texts.

What do you think?

File: adminservice/src/main/resources/templates/whitelisted-numbers.vm [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
+<html>
+<head><title>Whitelisted numbers</title></head>
```
Comment (Bert-R) [inline adminservice/src/main/resources/templates/whitelisted-numbers.vm]: White-listed, also for other (visible) occurrences on this page.

File: core/src/main/java/nu/yona/server/properties/YonaProperties.java [hunk 4]
```diff
@@ -45,6 +45,8 @@
 
 	private int maxUsers;
 
+	private boolean whitelistEnabled;
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/properties/YonaProperties.java]: We don't need this. Combining ``maxUsers`` with the white list feature is useless, so we can disable the white list feature when ``maxUsers`` is ``true``, and set ``maxUsers`` to a reasonably big number by default.

File: core/src/main/java/nu/yona/server/subscriptions/service/UserService.java [hunk 5]
```diff
@@ -230,6 +233,8 @@ private void verifyUserDoesNotExist(String mobileNumber)
 
 	private void verifyUserIsAllowed(String mobileNumber)
 	{
+		whitelistedNumberService.validateNumber(mobileNumber);
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/UserService.java]: See earlier comment on property. If you move this call down under ``if (maxUsers <= 0)``, it's taken care of.

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+	@Transactional
+	public Set<String> getAllWhitelistedNumbers()
+	{
+		return StreamSupport.stream(WhitelistedNumber.getRepository().findAll().spliterator(), false)
+				.map(whitelistedNumberEntity -> whitelistedNumberEntity.getMobileNumber()).collect(Collectors.toSet());
+	}
+
+	@Transactional
+	public void validateNumber(String mobileNumber)
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java]: Please rename into ``verifyMobileNumberIsAllowed``, in line with other method names. Methods starting with ``verify`` always throw when the verification fails.

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+
+	@Autowired
+	private YonaProperties yonaProperties;
+
+	@Transactional
+	public void addWhitelistedNumber(String mobileNumber)
+	{
+		userService.validateMobileNumber(mobileNumber);
+
+		WhitelistedNumber.getRepository().save(WhitelistedNumber.createInstance(mobileNumber));
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java]: It's more common and likely faster to autowire the repository.

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+				.map(whitelistedNumberEntity -> whitelistedNumberEntity.getMobileNumber()).collect(Collectors.toSet());
+	}
+
+	@Transactional
+	public void validateNumber(String mobileNumber)
+	{
+		if (!yonaProperties.getWhitelistEnabled())
+			return;
+
+		if (!getAllWhitelistedNumbers().contains(mobileNumber))
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java]: Why not ``if (whitelistedNumberRepository.findOne(mobileNumber) == null)``? That'll be faster than loading the full set.

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberServiceException.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+	}
+
+	private WhitelistedNumberServiceException(HttpStatus statusCode, String messageId, Object... parameters)
+	{
+		super(statusCode, messageId, parameters);
+	}
+
+	public static WhitelistedNumberServiceException numberNotWhitelisted(String mobileNumber)
+	{
+		return new WhitelistedNumberServiceException(HttpStatus.FORBIDDEN, "error.number.not.whitelisted", mobileNumber);
```
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberServiceException.java]: Isn't this error code too heavy? Users will also see this when inviting buddies that are not on the white list.

File: core/src/main/resources/messages/messages.properties [hunk 10]
```diff
@@ -33,6 +33,8 @@ error.user.exists.created.on.buddy.request=User with mobile number ''{0}'' alrea
 error.user.exists=User with mobile number ''{0}'' already exists
 error.user.maximum.number.reached=Reached the maximum number of users that is currently allowed
 
+error.number.not.whitelisted=Mobile number ''{0}'' does not occur in the whitelist of allowed numbers
```
Comment (Bert-R) [inline core/src/main/resources/messages/messages.properties]: Please change it this way: "Mobile number ''{0}'' does not occur on the white list of allowed numbers"

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberServiceException.java (hunk 9 above)
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberServiceException.java]: Changed to 400 now.

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java (hunk 8 above)
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java]: Done

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java (hunk 6 above)
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java]: Done

File: core/src/main/resources/messages/messages.properties (hunk 10 above)
Comment (aristotelos) [inline core/src/main/resources/messages/messages.properties]: Done

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java (hunk 7 above)
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java]: Currently it is not common in our code. I tried it after your suggestion, but get a bean not found error.

File: core/src/main/java/nu/yona/server/subscriptions/service/UserService.java (hunk 5 above)
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/subscriptions/service/UserService.java]: Done

File: core/src/main/java/nu/yona/server/properties/YonaProperties.java (hunk 4 above)
Comment (aristotelos) [inline core/src/main/java/nu/yona/server/properties/YonaProperties.java]: Done.

File: adminservice/src/main/java/nu/yona/server/subscriptions/rest/WhitelistedNumberController.java (hunk 2 above)
Comment (aristotelos) [inline adminservice/src/main/java/nu/yona/server/subscriptions/rest/WhitelistedNumberController.java]: Replaced all occurances with the US English variant.

File: adminservice/src/main/resources/templates/whitelisted-numbers.vm (hunk 1 above)
Comment (aristotelos) [inline adminservice/src/main/resources/templates/whitelisted-numbers.vm]: Done

File: adminservice/src/main/resources/templates/whitelisted-numbers.vm (hunk 3 above)
Comment (aristotelos) [inline adminservice/src/main/resources/templates/whitelisted-numbers.vm]: Done

File: adminservice/src/main/java/nu/yona/server/subscriptions/rest/WhitelistedNumberController.java (hunk 2 above)
Comment (Bert-R) [inline adminservice/src/main/java/nu/yona/server/subscriptions/rest/WhitelistedNumberController.java]: Super!

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java (hunk 7 above)
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberService.java]: Indeed, in our code it is, but our code is special. We can stick with it for now. I'm taking out this pattern where I touch it.

File: core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberServiceException.java (hunk 9 above)
Comment (Bert-R) [inline core/src/main/java/nu/yona/server/subscriptions/service/WhitelistedNumberServiceException.java]: OK

**Pull request merged**